_posts/2019-08-08-kubernetes-gated-deployments.md

```diff
@@ -0,0 +1,245 @@
+* `FAIL`: if the treatment does harm to the metric analyzed
+
+Teams can configure multiple plugins for a single deployment. For example, a
+team might want to measure response time and error rate for multiple endpoints
+in a single application along with. In this case, all plugins must return `PASS`
+for the controller to decide that the experiment is a success; if any plugin
+returns `FAIL`, it will decide that the experiment is a failure. Otherwise, the
+controller will continue polling the metrics until the plugin satisfies one of
+the two conditions above. Once the controller makes a decision for the
+experiment and takes action to roll back or continue deploying the treatment,
+the controller will set an annotation, `gatedDeployStatus`, on the treatment
+`Deployment` to specify what the decision was. Teams can use this to observe the
+outcome of the gated deployment or in CICD pipelines to programmatically halt
+the deployment process on a failed experiment.
+
+For a list of the decision plugins available and to contribute more, see
+[here](https://github.com/godaddy/kubernetes-gated-deployments/tree/master/lib/plugins).
+
+## Usage of the New Relic decision plugin
+
+A decision plugin that we implemented and have been using at GoDaddy is a
+`newRelicPerformance` plugin, which analyzes response times to determine if the
+newly deployed version is performing worse than the current version. Shown below
+is a sample configuration of the plugin in the `GatedDeployment` object:
+
+```yaml
+apiVersion: 'kubernetes-client.io/v1'
+kind: GatedDeployment
+metadata:
+  name: example-rest-service
+deploymentDescriptor:
+  control:
+    name: example-rest-service-control
+  treatment:
+    name: example-rest-service-treatment
+  decisionPlugins:
+    - name: newRelicPerformance
+      accountId: 807783
+      secretName: newrelic-secrets
+      secretKey: example-rest-service
+      appName: example-rest-service
+      minSamples: 50
+      maxTime: 600
+      testPath: /shopper/products
+```
+
+This plugin retrieves response times from New Relic for the control and
+treatment versions, and tests if they are statistically significantly different
+using the [Mann-Whitney U
+test](https://en.wikipedia.org/wiki/Mann–Whitney_U_test). Because statistical
+significance does not necessarily imply practical significance, the plugin
+exposes a threshold setting before it marks the treatment version as causing
+harm to the metric of response time. This is useful in the case when the
+application does not receive much traffic, as small sample sizes can result in
+erroneous statistically significant results. Depending on the traffic patterns
+and response time distribution of the application, teams can also configure a
+minimum number of samples before analyzing the response times, and a maximum
+amount of time in seconds that the experiment will run for before automatically
+deciding that it is a success.
+
+## Kubernetes Gated Deployments in production
```

Comment:
It'd be nice to see a visualization of what pass/fail looks like here - what does New Relic show for a positive case and a negative case?